gdal/alg/gdalrasterize.cpp

```diff
@@ -670,90 +703,260 @@ CPLErr GDALRasterizeGeometries( GDALDatasetH hDS,
+        }
+
+        int * panSuccessTransform = (int *) CPLCalloc(sizeof(int), 2);
+        
+/* -------------------------------------------------------------------- */
+/*      loop over the vectorial geometries                              */
+/* -------------------------------------------------------------------- */
+        pfnProgress( 0.0, NULL, pProgressArg );
         for( int iShape = 0; iShape < nGeomCount; iShape++ )
         {
-            gv_rasterize_one_shape( pabyChunkBuf, iY,
-                                    poDS->GetRasterXSize(), nThisYChunkSize,
-                                    nBandCount, eType, bAllTouched,
-                                    reinterpret_cast<OGRGeometry *>(
-                                                        pahGeometries[iShape]),
-                                    padfGeomBurnValue + iShape*nBandCount,
-                                    eBurnValueSource, eMergeAlg,
-                                    pfnTransformer, pTransformArg );
+
+/* -------------------------------------------------------------------- */
+/*      get the envelope of the geometry and transform it to pixels coo */
+/* -------------------------------------------------------------------- */
+            OGREnvelope psGeomEnvelope;
+            ((OGRGeometry *) pahGeometries[iShape])->getEnvelope(&psGeomEnvelope);
+            if( pfnTransformer != NULL )
+            {
+                double apCorners[4];
+                apCorners[0] = psGeomEnvelope.MinX;
+                apCorners[1] = psGeomEnvelope.MaxX;
+                apCorners[2] = psGeomEnvelope.MinY;
+                apCorners[3] = psGeomEnvelope.MaxY;
+                // TODO: needad to add all appropriate error checking 
+                pfnTransformer( pTransformArg, FALSE, 2, &(apCorners[0]),
+                                &(apCorners[2]), NULL, panSuccessTransform );
+                psGeomEnvelope.MinX = std::min(apCorners[0], apCorners[1]);
+                psGeomEnvelope.MaxX = std::max(apCorners[0], apCorners[1]);
+                psGeomEnvelope.MinY = std::min(apCorners[2], apCorners[3]);
+                psGeomEnvelope.MaxY = std::max(apCorners[2], apCorners[3]);
+            }
+
+
+            int minBlockX = std::max(0, int(psGeomEnvelope.MinX) / nXBlockSize );
+            int minBlockY = std::max(0, int(psGeomEnvelope.MinY) / nYBlockSize );
+            int maxBlockX = std::min(nXBlocks-1, int(psGeomEnvelope.MaxX+1) / nXBlockSize );
+            int maxBlockY = std::min(nYBlocks-1, int(psGeomEnvelope.MaxY+1) / nYBlockSize );
+
+            
+
+/* -------------------------------------------------------------------- */
+/*      loop over the blocks concerned by the geometry                  */
+/*      (by packs of nbBlocsX x nbBlocsY)                                 */
+/* -------------------------------------------------------------------- */
+
+            for(int xB = minBlockX; xB <= maxBlockX; xB += nbBlocsX)
```

Comment:
Instead of looping and doing several read, rasterize, write, wouldn't it be faster to do that in a single step ? You've already allocated the pabyChunkBuf buffer to be large enough.